pkg/agent/informant.go

```diff
@@ -413,13 +436,70 @@ func (s *InformantServer) RegisterWithInformant(ctx context.Context, logger *zap
 	resp, statusCode, err := doInformantRequest[api.AgentDesc, api.InformantDesc](
 		ctx, logger, s, timeout, http.MethodPost, "/register", &s.desc,
 	)
+
+	connectedToMonitor := false
+
 	// Do some stuff with the lock acquired:
 	func() {
 		maybeMadeContact := statusCode != 0 || ctx.Err() != nil
 
 		s.runner.lock.Lock()
 		defer s.runner.lock.Unlock()
 
+		if statusCode == 404 {
+			addr := fmt.Sprintf(
+				"ws://%s:%d/monitor",
+				s.runner.podIP,
+				s.runner.global.config.Informant.ServerPort,
+			)
+			logger.Info(
+				"received 404 from informant's /register endpoint; connecting to monitor",
+				zap.String("addr", addr),
+			)
+			// pre-declare disp so that err get's assigned to err from enclosing scope,
+			// overwriting original request error.
+			var disp Dispatcher
+			disp, err = NewDispatcher(logger, addr, s)
```

Comment:
Does `NewDispatcher` make a network request?

If so, we shouldn't block `s.runner.lock` while holding it